source/common/http/utility.cc

```diff
@@ -281,37 +282,57 @@ Utility::parseHttp1Settings(const envoy::api::v2::core::Http1ProtocolOptions& co
   return ret;
 }
 
-void Utility::sendLocalReply(bool is_grpc, StreamDecoderFilterCallbacks& callbacks,
-                             const bool& is_reset, Code response_code, absl::string_view body_text,
-                             const absl::optional<Grpc::Status::GrpcStatus> grpc_status,
-                             bool is_head_request) {
-  sendLocalReply(is_grpc,
+Utility::LocalReplyInfo Utility::generateLocalReplyInfo(const Http::HeaderMap& request_headers) {
+  bool is_grpc = false, is_head_request = false, accept_json_type = false;
+
+  is_grpc = Grpc::Utility::hasGrpcContentType(request_headers);
+
+  if (Http::Headers::get().MethodValues.Head == request_headers.Method()->value().c_str()) {
+    is_head_request = true;
+  }
+
+  if (!is_grpc) {
+    const Http::HeaderEntry* accept = request_headers.Accept();
+    // As long as can accept the json type response priority return json type.
+    accept_json_type = accept != nullptr &&
+                       (accept->value().find(Http::Headers::get().ContentTypeValues.Json.c_str()));
```

Comment:
I think we need to support full q-value parsing here if we are going to handle this. @jmarantz do you mind taking a look at the this PR in general and seeing what you think in addition to commenting on this? I'm still wondering if we should make this behavior configurable, although it seems like a great enhancement to me overall.

Reply:
FYI I needed to look this up myself, and this looks like a good website to get an overview; https://www.newmediacampaigns.com/blog/browser-rest-http-accept-headers and also https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation

It does seem like a broader framework is desired ultimately to fully support that spec, rather than special-casing grpc and json here.

It might be better to have this configurable as it looks like there's more we can/should do, and we probably don't want to boil the ocean before supporting json responses which is the more modest goal of this PR.

Maybe consider this an experimental feature short term and just flag-protect until a more complete design can be proposed?

Reply:
One more note, the envoy gzip filter already has q-value parsing in it, which probably should be factored out rather than reimplemented out copied.

Reply:
One more note, the envoy gzip filter already has q-value parsing in it, which probably should be factored out rather than reimplemented out copied.

Reply:
One more note, the envoy gzip filter already has q-value parsing in it, which probably should be factored out rather than reimplemented out copied.